File: clients/js/src/sandbox.ts
/**
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import WebSocket from 'ws';
import { EventEmitter } from 'events';
import { MessageKey, EventType, SandboxEvent, WebSocketMessage } from './types';
import { SandboxProcess } from './process';
import { Connection, ShouldReconnectCallback, ReconnectInfo } from './connection';

type SandboxState = 'creating' | 'running' | 'closed' | 'failed' | 'checkpointing' | 'checkpointed' | 'restoring' | 'filesystem_snapshotting' | 'reconnecting';

export class Sandbox {
  private connection: Connection;
  private eventEmitter = new EventEmitter();
  private _sandboxId: string | null = null;
  private _sandboxToken: string | null = null;
  private _creationError: Error | null = null;
  private activeProcess: SandboxProcess | null = null;
  private state: SandboxState = 'creating';
  private _debugEnabled: boolean = false;
  private _debugLabel: string = '';
  private _shouldReconnect: boolean = false;
  private _autoReconnectEnabled: boolean = false;
  private _isCheckpointIntentionally: boolean = false;
  private _isKillIntentionally = false;
  private _url: string = '';
  private _wsOptions?: WebSocket.ClientOptions;
  private stdinBuffer: string[] = [];

  private constructor(connection: Connection, debug: boolean = false, debugLabel: string = '', autoReconnectEnabled: boolean = false) {
    this.connection = connection;
    this._debugEnabled = debug;
    this._debugLabel = debugLabel;
    this._autoReconnectEnabled = autoReconnectEnabled;
    this.connection.on('message', this.handleMessage.bind(this));
    this.connection.on('close', this.handleClose.bind(this));
    this.connection.on('error', this.handleError.bind(this));
    this.connection.on('reopen', this.handleReopen.bind(this));
  }

  public get sandboxId(): string | null {
    return this._sandboxId;
  }

  public get sandboxToken(): string | null {
    return this._sandboxToken;
  }

  private logDebugMessage(message: string, ...args: any[]) {
    if (this._debugEnabled) {
      console.log(`[${this._debugLabel}] [DEBUG] ${message}`, ...args);
    }
  }

  private _updateShouldReconnect(status: SandboxEvent) {
    const isFatalError = [
      SandboxEvent.SANDBOX_ERROR,
      SandboxEvent.SANDBOX_NOT_FOUND,
      SandboxEvent.SANDBOX_CREATION_ERROR,
      SandboxEvent.SANDBOX_CHECKPOINT_ERROR,
      SandboxEvent.SANDBOX_RESTORE_ERROR,
      SandboxEvent.SANDBOX_DELETED,
      SandboxEvent.SANDBOX_LOCK_RENEWAL_ERROR,
      SandboxEvent.SANDBOX_PERMISSION_DENIAL_ERROR,
    ].includes(status);

    if (isFatalError) {
      this._shouldReconnect = false;
    } else if (status === SandboxEvent.SANDBOX_RUNNING) {
      this._shouldReconnect = this._autoReconnectEnabled;
    }
  }

  private handleMessage(data: WebSocket.Data) {
    const message: WebSocketMessage = JSON.parse(data.toString());

    if (message.event != EventType.STDOUT && message.event != EventType.STDERR) {
      this.logDebugMessage('Received message:', message);
    }

    // Process-specific events are always forwarded
    if (
      message.event === EventType.STDOUT ||
      message.event === EventType.STDERR ||
      (message.event === EventType.STATUS_UPDATE &&
        message.status &&
        message.status.startsWith('SANDBOX_EXECUTION_') &&
        message.status !== SandboxEvent.SANDBOX_EXECUTION_IN_PROGRESS_ERROR)
    ) {
      if (this.activeProcess) {
        this.activeProcess.handleMessage(message);
      }
      return;
    }

    // Handle sandbox lifecycle events
    if (message.event === EventType.SANDBOX_ID) {
      this._sandboxId = message.sandbox_id;
      this._sandboxToken = message.sandbox_token ?? null;
      return;
    }

    if (message.event === EventType.STATUS_UPDATE) {
      this._updateShouldReconnect(message.status);

      switch (message.status) {
        case SandboxEvent.SANDBOX_RUNNING:
          if (this.state === 'reconnecting') {
            this.flushStdinBuffer();
          }
          this.state = 'running';
          this.eventEmitter.emit('created', this);
          break;
        case SandboxEvent.SANDBOX_CREATION_ERROR:
          if (this.state === 'creating') {
            this.state = 'failed';
            this._creationError = new Error(message.message || message.status);
            this.eventEmitter.emit('failed', this._creationError);
          }
          break;
        case SandboxEvent.SANDBOX_PERMISSION_DENIAL_ERROR:
          this.state = 'failed';
          this._creationError = new Error(message.message || message.status);
          this.eventEmitter.emit('failed', this._creationError);
          break;
        case SandboxEvent.SANDBOX_CHECKPOINTING:
          this.state = 'checkpointing';
          break;
        case SandboxEvent.SANDBOX_CHECKPOINTED:
          this.state = 'checkpointed';
          this.eventEmitter.emit('checkpointed');
          break;
        case SandboxEvent.SANDBOX_CHECKPOINT_ERROR:
          this.state = 'failed'; // This is a fatal error for the session.
          this.eventEmitter.emit('checkpoint_error', new Error(message.message || message.status));
          break;
        case SandboxEvent.SANDBOX_EXECUTION_IN_PROGRESS_ERROR:
          this.state = 'running'; // Checkpoint failed, back to running
          this.eventEmitter.emit('checkpoint_error', new Error(message.message || message.status));
          break;
        case SandboxEvent.SANDBOX_RESTORING:
          this.state = 'restoring';
          break;
        case SandboxEvent.SANDBOX_RESTORE_ERROR:
          this.state = 'failed';
          this._creationError = new Error(message.message || message.status);
          this.eventEmitter.emit('failed', this._creationError);
          break;
        case SandboxEvent.SANDBOX_NOT_FOUND:
          this.state = 'failed';
          this._creationError = new Error(message.message || message.status);
          this.eventEmitter.emit('failed', this._creationError);
          break;
        case SandboxEvent.SANDBOX_FILESYSTEM_SNAPSHOT_CREATING:
          this.state = 'filesystem_snapshotting';
          break;
        case SandboxEvent.SANDBOX_FILESYSTEM_SNAPSHOT_CREATED:
          this.state = 'running';
          this.eventEmitter.emit('filesystem_snapshot_created');
          break;
        case SandboxEvent.SANDBOX_FILESYSTEM_SNAPSHOT_ERROR:
          this.state = 'running';
          this.eventEmitter.emit('filesystem_snapshot_error', new Error(message.message || message.status));
          break;
        case SandboxEvent.SANDBOX_KILLED:
        case SandboxEvent.SANDBOX_KILL_ERROR:
          this.eventEmitter.emit('killed');
          break;
      }
      return;
    }
  }

  private handleClose(code: number, reason: Buffer) {
    this.logDebugMessage(`Connection closed: code=${code}, reason=${reason ? reason.toString() : 'No reason'}`);
    if (this.state === 'creating' || this.state === 'restoring') {
      this.state = 'failed';
      const err = new Error(`Connection closed during creation/restoration: code=${code}`);
      this.eventEmitter.emit('failed', err);
    } else {
      this.state = 'closed';
    }
    
    if (this.activeProcess) {
      this.activeProcess.close();
      this.activeProcess = null;
    }
  }

  private handleError(err: Error) {
    if (this._debugEnabled) {
      console.error(`[${this._debugLabel}] [DEBUG] WebSocket error:`, err);
    }
    if (this.state === 'creating' || this.state === 'restoring') {
      this.state = 'failed';
      this.eventEmitter.emit('failed', err);
    }
    
    if (this.activeProcess) {
      this.activeProcess.close();
      this.activeProcess = null;
    }
    // In the 'running' state, you might want to emit a general error
    // for the user to handle, e.g., this.eventEmitter.emit('error', err);
  }

  private handleReopen() {
    this.logDebugMessage('Reconnected. Sending reconnect action.');
    this.connection.send(JSON.stringify({ action: 'reconnect' }));
  }

  public shouldReconnect(code: number, reason: Buffer): boolean {
    const decision = this._shouldReconnect && !this._isCheckpointIntentionally;
    this.logDebugMessage(`Checking if should reconnect: code=${code}, reason=${reason.toString()}, decision=${decision}`);
    if (decision) {
      this.state = 'reconnecting';
    }
    return decision;
  }

  private getReconnectInfo(): ReconnectInfo {
    if (!this._sandboxId) {
      throw new Error('Cannot reconnect without a sandbox ID.');
    }
    const sanitizedUrl = this._url.replace(/\/$/, '');
    const reconnectUrl = `${sanitizedUrl}/attach/${this._sandboxId}?sandbox_token=${this._sandboxToken}`;
    return { url: reconnectUrl, wsOptions: this._wsOptions };
  }

  static create(url: string, options: { idleTimeout?: number, enableSandboxCheckpoint?: boolean, enableSandboxHandoff?: boolean, filesystemSnapshotName?: string, enableDebug?: boolean, debugLabel?: string, wsOptions?: WebSocket.ClientOptions, enableAutoReconnect?: boolean, enableIdleTimeoutAutoCheckpoint?: boolean } = {}): Promise<Sandbox> {
    const { idleTimeout = 60, enableSandboxCheckpoint = false, enableSandboxHandoff = false, filesystemSnapshotName, enableDebug = false, debugLabel = '', wsOptions, enableAutoReconnect = false, enableIdleTimeoutAutoCheckpoint = false } = options;
    
    const sanitizedUrl = url.replace(/\/$/, '');
    let sandbox: Sandbox;
    const connection = new Connection(
      `${sanitizedUrl}/create`,
      (code, reason) => sandbox.shouldReconnect(code, reason),
      () => sandbox.getReconnectInfo(),
      wsOptions,
      enableDebug,
      debugLabel,
    );
    sandbox = new Sandbox(connection, enableDebug, debugLabel, enableAutoReconnect);
    sandbox._url = url;
    sandbox._wsOptions = wsOptions;

    connection.once('open', () => {
      connection.send(JSON.stringify({
        idle_timeout: idleTimeout,
        enable_checkpoint: enableSandboxCheckpoint,
        enable_sandbox_handoff: enableSandboxHandoff,
        filesystem_snapshot_name: filesystemSnapshotName,
        enable_idle_timeout_auto_checkpoint: enableIdleTimeoutAutoCheckpoint,
      }));
    });
    
    return new Promise((resolve, reject) => {
      sandbox.eventEmitter.once('created', (createdSandbox) => {
        resolve(createdSandbox);
      });
      
      sandbox.eventEmitter.once('failed', (err) => {
        // On failure, ensure the socket is completely destroyed.
        connection.close();
        reject(err);
      });
    });
  }

  static attach(url: string, sandboxId: string, sandboxToken: string, options: { enableDebug?: boolean, debugLabel?: string, wsOptions?: WebSocket.ClientOptions, enableAutoReconnect?: boolean } = {}): Promise<Sandbox> {
    const { enableDebug = false, debugLabel = '', wsOptions, enableAutoReconnect = false } = options;
    
    const sanitizedUrl = url.replace(/\/$/, '');
    let sandbox: Sandbox;
    const connection = new Connection(
      `${sanitizedUrl}/attach/${sandboxId}?sandbox_token=${sandboxToken}`,
      (code, reason) => sandbox.shouldReconnect(code, reason),
      () => sandbox.getReconnectInfo(),
      wsOptions,
      enableDebug,
      debugLabel,
    );
    sandbox = new Sandbox(connection, enableDebug, debugLabel, enableAutoReconnect);
    sandbox._url = url;
    sandbox._wsOptions = wsOptions;
    sandbox._sandboxId = sandboxId;
    sandbox._sandboxToken = sandboxToken;

    return new Promise((resolve, reject) => {
      sandbox.eventEmitter.once('created', (createdSandbox) => {
        resolve(createdSandbox);
      });
      
      sandbox.eventEmitter.once('failed', (err) => {
        // On failure, ensure the socket is completely destroyed.
        connection.close();
        reject(err);
      });
    });
  }

  public checkpoint(): Promise<void> {
    if (this.state !== 'running') {
      return Promise.reject(new Error(`Sandbox is not in a running state. Current state: ${this.state}`));
    }
    this._isCheckpointIntentionally = true;
    this.connection.send(JSON.stringify({ action: 'checkpoint' }));

    return new Promise((resolve, reject) => {
      this.eventEmitter.once('checkpointed', () => {
        resolve();
      });
      this.eventEmitter.once('checkpoint_error', (err) => {
        reject(err);
      });
    });
  }

  public snapshotFilesystem(name: string): Promise<void> {
    if (this.state !== 'running') {
      return Promise.reject(new Error(`Sandbox is not in a running state. Current state: ${this.state}`));
    }

    this.connection.send(JSON.stringify({ action: 'snapshot_filesystem', name }));

    return new Promise((resolve, reject) => {
      this.eventEmitter.once('filesystem_snapshot_created', () => {
        resolve();
      });
      this.eventEmitter.once('filesystem_snapshot_error', (err) => {
        reject(err);
      });
    });
  }

  private flushStdinBuffer() {
    this.logDebugMessage(`Flushing stdin buffer (${this.stdinBuffer.length} messages).`);
    for (const data of this.stdinBuffer) {
      this.connection.send(data);
    }
    this.stdinBuffer = [];
  }

  private sendMessage(data: string) {
    if (this.state === 'reconnecting') {
      try {
        const message = JSON.parse(data);
        if (message.event === 'stdin') {
          this.logDebugMessage('Buffering stdin message while reconnecting:', message.data);
          this.stdinBuffer.push(data);
          return;
        }
      } catch (e) {
        // Not a JSON message, or doesn't have an event property.
        // Let it pass through. This shouldn't happen for stdin.
      }
    }
    this.connection.send(data);
  }

  public async exec(language: string, code: string): Promise<SandboxProcess> {
    if (this.activeProcess) {
      throw new Error('Another process is already running in this sandbox.');
    }
    if (this.state !== 'running') {
      throw new Error(`Sandbox is not in a running state. Current state: ${this.state}`);
    }

    const process = new SandboxProcess(
      this.sendMessage.bind(this),
    );
    this.activeProcess = process;
    
    process['eventEmitter'].once('done', () => {
      this.activeProcess = null;
    });

    await process.exec(language, code);
    return process;
  }

  public kill(): Promise<void> {
    if (this.state === 'closed' || this.state === 'failed' || this.state === 'checkpointed') {
      this.connection.close();
      return Promise.resolve();
    }

    return new Promise((resolve) => {
      const timeout = setTimeout(() => {
        this.logDebugMessage('Kill timeout reached. Forcing connection close.');
        this.connection.close();
        resolve();
      }, 5000); // 5-second timeout

      this.eventEmitter.once('killed', () => {
        this.logDebugMessage('Sandbox killed successfully.');
        clearTimeout(timeout);
        this.connection.close();
        resolve();
      });

      this._isKillIntentionally = true;
      this.logDebugMessage('Sending kill command to sandbox...');
      this.connection.send(JSON.stringify({
        action: 'kill_sandbox',
      }));
    });
  }
}

File: clients/python/src/sandbox/sandbox.py

# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import asyncio
import json
import weakref
import websockets

from .types import MessageKey, EventType, SandboxEvent
from .exceptions import SandboxConnectionError, SandboxCreationError, SandboxStateError, SandboxFilesystemSnapshotError, SandboxCheckpointError
from .process import SandboxProcess
from .connection import Connection

class Sandbox:
    """
    Represents a connection to a Cloud Run Sandbox, used to execute commands.
    """
    def __init__(self, initial_state: str, enable_debug=False, debug_label=""):
        self._connection = None
        self._sandbox_id = None
        self._sandbox_token = None
        self._active_process = None
        self._ready_event = asyncio.Event()
        self._killed_event = asyncio.Event()
        self._filesystem_snapshot_event = asyncio.Event()
        self._filesystem_snapshot_error = None
        self._checkpoint_event = asyncio.Event()
        self._checkpoint_error = None
        self._error_on_ready = None
        self._state = initial_state
        self._shutdown_lock = asyncio.Lock()
        self._enable_debug = enable_debug
        self._debug_label = debug_label
        self._is_kill_intentionally = False
        self._enable_auto_reconnect = False
        self._should_reconnect_internal = False
        self._base_url = None

    @property
    def sandbox_id(self):
        """The unique ID of the sandbox."""
        return self._sandbox_id

    @property
    def sandbox_token(self):
        """The token for the sandbox."""
        return self._sandbox_token

    @classmethod
    async def create(cls, url: str, idle_timeout: int = 60, ssl=None, enable_debug=False, debug_label="", filesystem_snapshot_name: str = None, enable_sandbox_checkpoint: bool = False, enable_idle_timeout_auto_checkpoint: bool = False, enable_auto_reconnect: bool = False, enable_sandbox_handoff: bool = False):
        """
        Creates a new sandbox session.
        """
        instance = cls("creating", enable_debug, debug_label)
        instance._enable_auto_reconnect = enable_auto_reconnect
        sanitized_url = url.rstrip('/')
        instance._base_url = sanitized_url
        ws_url = f"{sanitized_url}/create"
        
        ws_options = {"ssl": ssl} if ssl else {}

        try:
            instance._connection = Connection(
                url=ws_url,
                on_message=instance._on_message,
                on_error=instance._on_error,
                on_close=instance._on_close,
                should_reconnect=instance._should_reconnect,
                get_reconnect_info=instance._get_reconnect_info,
                on_reopen=instance._on_reopen,
                ws_options=ws_options,
                debug=enable_debug,
                debug_label=debug_label,
            )
            await instance._connection.connect()
        except Exception as e:
            raise SandboxConnectionError(f"Failed to connect to {ws_url}: {e}")

        create_params = {"idle_timeout": idle_timeout}
        if filesystem_snapshot_name:
            create_params["filesystem_snapshot_name"] = filesystem_snapshot_name
        if enable_sandbox_checkpoint:
            create_params["enable_checkpoint"] = enable_sandbox_checkpoint
        if enable_idle_timeout_auto_checkpoint:
            create_params["enable_idle_timeout_auto_checkpoint"] = enable_idle_timeout_auto_checkpoint
        if enable_sandbox_handoff:
            create_params["enable_sandbox_handoff"] = enable_sandbox_handoff
        await instance._send(create_params)
        
        await instance._wait_for_ready()
        return instance

    @classmethod
    async def attach(cls, url: str, sandbox_id: str, sandbox_token: str, ssl=None, enable_debug=False, debug_label="", enable_auto_reconnect: bool = False):
        """
        Attaches to an existing sandbox session.
        """
        instance = cls("attaching", enable_debug, debug_label)
        instance._enable_auto_reconnect = enable_auto_reconnect
        instance._sandbox_id = sandbox_id
        instance._sandbox_token = sandbox_token
        sanitized_url = url.rstrip('/')
        instance._base_url = sanitized_url
        ws_url = f"{sanitized_url}/attach/{sandbox_id}?sandbox_token={sandbox_token}"

        ws_options = {"ssl": ssl} if ssl else {}

        try:
            instance._connection = Connection(
                url=ws_url,
                on_message=instance._on_message,
                on_error=instance._on_error,
                on_close=instance._on_close,
                should_reconnect=instance._should_reconnect,
                get_reconnect_info=instance._get_reconnect_info,
                on_reopen=instance._on_reopen,
                ws_options=ws_options,
                debug=enable_debug,
                debug_label=debug_label,
            )
            await instance._connection.connect()
        except Exception as e:
            raise SandboxConnectionError(f"Failed to connect to {ws_url}: {e}")

        await instance._wait_for_ready()
        return instance

    def _log_debug(self, message):
        if self._enable_debug:
            label = f"[{self._debug_label}]" if self._debug_label else ""
            print(f"{label} {message}")

    async def _send(self, data: dict):
        """Sends a JSON message to the WebSocket."""
        message_str = json.dumps(data)
        if self._connection:
            await self._connection.send(message_str)

    async def _wait_for_ready(self):
        """
        Waits for the initial handshake to complete and the sandbox to be running.
        """
        await self._ready_event.wait()
        if self._error_on_ready:
            raise self._error_on_ready

    def _update_should_reconnect(self, status: SandboxEvent):
        is_fatal_error = status in [
            SandboxEvent.SANDBOX_ERROR,
            SandboxEvent.SANDBOX_NOT_FOUND,
            SandboxEvent.SANDBOX_CREATION_ERROR,
            SandboxEvent.SANDBOX_CHECKPOINT_ERROR,
            SandboxEvent.SANDBOX_RESTORE_ERROR,
            SandboxEvent.SANDBOX_DELETED,
            SandboxEvent.SANDBOX_LOCK_RENEWAL_ERROR,
            SandboxEvent.SANDBOX_PERMISSION_DENIAL_ERROR,
        ]

        if is_fatal_error:
            self._should_reconnect_internal = False
        elif status == SandboxEvent.SANDBOX_RUNNING:
            self._should_reconnect_internal = self._enable_auto_reconnect

    def _on_message(self, message_str: str):
        """Callback for handling incoming WebSocket messages."""
        message = json.loads(message_str)
        event = message.get(MessageKey.EVENT)

        # Handle noisy I/O events first and skip logging for them.
        if event in [EventType.STDOUT, EventType.STDERR]:
            if self._active_process:
                self._active_process.handle_message(message)
            return

        # For all other messages, log first.
        self._log_debug(f"Received message: {message_str}")

        # Now, dispatch the message.
        is_execution_status = (event == EventType.STATUS_UPDATE and
                               message.get(MessageKey.STATUS, "").startswith("SANDBOX_EXECUTION_") and
                               message.get(MessageKey.STATUS) !=  SandboxEvent.SANDBOX_EXECUTION_IN_PROGRESS_ERROR)

        if is_execution_status:
            if self._active_process:
                self._active_process.handle_message(message)
            return
        
        # Handle sandbox lifecycle events
        if event == EventType.SANDBOX_ID:
            self._sandbox_id = message.get(MessageKey.SANDBOX_ID)
            self._sandbox_token = message.get(MessageKey.SANDBOX_TOKEN)
            return

        if event == EventType.STATUS_UPDATE:
            status = message.get(MessageKey.STATUS)
            self._update_should_reconnect(status)
            if status == SandboxEvent.SANDBOX_RUNNING:
                self._state = "running"
                self._ready_event.set()
            elif status == SandboxEvent.SANDBOX_RESTORING:
                self._state = "restoring"
            elif status in [SandboxEvent.SANDBOX_CREATION_ERROR, SandboxEvent.SANDBOX_NOT_FOUND, SandboxEvent.SANDBOX_IN_USE, SandboxEvent.SANDBOX_RESTORE_ERROR, SandboxEvent.SANDBOX_PERMISSION_DENIAL_ERROR]:
                error = SandboxCreationError(message.get(MessageKey.MESSAGE, status))
                self._handle_error_on_ready(error)
                asyncio.create_task(self._shutdown())
            elif status in [SandboxEvent.SANDBOX_KILLED, SandboxEvent.SANDBOX_KILL_ERROR]:
                self._killed_event.set()
            elif status == SandboxEvent.SANDBOX_FILESYSTEM_SNAPSHOT_CREATED:
                self._state = "running"
                self._filesystem_snapshot_event.set()
            elif status == SandboxEvent.SANDBOX_FILESYSTEM_SNAPSHOT_ERROR:
                self._filesystem_snapshot_error = SandboxFilesystemSnapshotError(message.get(MessageKey.MESSAGE, status))
                self._filesystem_snapshot_event.set()
            elif status == SandboxEvent.SANDBOX_CHECKPOINTING:
                self._state = "checkpointing"
            elif status == SandboxEvent.SANDBOX_CHECKPOINTED:
                self._state = "checkpointed"
                self._checkpoint_event.set()
            elif status == SandboxEvent.SANDBOX_CHECKPOINT_ERROR:
                self._checkpoint_error = SandboxCheckpointError(message.get(MessageKey.MESSAGE, status))
                self._checkpoint_event.set()
            elif status == SandboxEvent.SANDBOX_EXECUTION_IN_PROGRESS_ERROR:
                self._state = "running"
                self._checkpoint_error = SandboxCheckpointError(message.get(MessageKey.MESSAGE, status))
                self._checkpoint_event.set()

    def _on_error(self, error: Exception):
        """Callback for handling WebSocket connection errors."""
        self._log_debug(f"An unexpected error occurred: {error}")
        self._handle_error_on_ready(error)
        asyncio.create_task(self._shutdown())

    def _on_close(self, code: int, reason: str):
        """Callback for when the WebSocket connection is closed."""
        error = SandboxConnectionError(f"Connection closed: {code} - {reason}")
        self._log_debug(f"Connection closed: {code} - {reason}")
        self._handle_error_on_ready(error)
        asyncio.create_task(self._shutdown())

    def _should_reconnect(self, code: int, reason: str) -> bool:
        return self._should_reconnect_internal and not self._is_kill_intentionally

    def _get_reconnect_info(self):
        base_url = self._base_url
        reconnect_url = f"{base_url}/attach/{self._sandbox_id}?sandbox_token={self._sandbox_token}"
        return {"url": reconnect_url, "ws_options": self._connection.ws_options}

    async def _on_reopen(self):
        self._log_debug("Reconnected. Sending reconnect action.")
        await self._send({"action": "reconnect"})

    async def snapshot_filesystem(self, name: str):
        """
        Triggers the creation of a filesystem snapshot.
        """
        if self._state != "running":
            raise SandboxStateError(f"Sandbox is not in a running state. Current state: {self._state}")

        self._log_debug(f"Requesting filesystem snapshot '{name}'...")
        self._state = "filesystem_snapshotting"
        self._filesystem_snapshot_event.clear()
        self._filesystem_snapshot_error = None

        await self._send({"action": "snapshot_filesystem", "name": name})
        await self._filesystem_snapshot_event.wait()
        if self._filesystem_snapshot_error:
            raise self._filesystem_snapshot_error

    async def checkpoint(self):
        """
        Triggers the creation of a checkpoint.
        """
        if self._state != "running":
            raise SandboxStateError(f"Sandbox is not in a running state. Current state: {self._state}")

        self._log_debug("Requesting checkpoint...")
        self._state = "checkpointing"
        self._checkpoint_event.clear()
        self._checkpoint_error = None

        await self._send({"action": "checkpoint"})
        await self._checkpoint_event.wait()
        if self._checkpoint_error:
            raise self._checkpoint_error
        await self._shutdown()

    async def exec(self, language: str, code: str) -> SandboxProcess:
        """
        Creates and starts a new process in the sandbox for code execution.
        """
        if self._active_process:
            raise RuntimeError("Another process is already running in this sandbox.")
        if self._state != "running":
            raise SandboxStateError(f"Sandbox is not in a running state. Current state: {self._state}")

        process = SandboxProcess(self._send, on_done=self._clear_active_process)
        self._active_process = process
        
        await process.exec(language, code)
        return process

    def _clear_active_process(self):
        self._active_process = None

    def _handle_error_on_ready(self, exc):
        if self._state in ["creating", "attaching"]:
            self._state = "failed"
            self._error_on_ready = exc
            self._ready_event.set()
            self._log_debug(f"Initialization failed: {exc}")

    async def _shutdown(self):
        async with self._shutdown_lock:
            if self._state in ["closed", "failed"]:
                return

            self._state = "closed"
            self._log_debug("State changed to closed")

            if self._active_process:
                await self._active_process.kill()
                self._active_process = None

            if self._connection:
                await self._connection.close()

    async def kill(self, timeout: float = 5.0):
        """
        Terminates the sandbox session.
        """
        if self._state in ["closed", "failed", "checkpointed"]:
            return

        self._is_kill_intentionally = True
        self._log_debug("Sending kill command to sandbox...")
        await self._send({"action": "kill_sandbox"})

        try:
            await asyncio.wait_for(self._killed_event.wait(), timeout=timeout)
            self._log_debug("Sandbox killed successfully.")
        except asyncio.TimeoutError:
            self._log_debug("Kill timeout reached. Forcing connection close.")
        finally:
            await self._shutdown()

    async def __aenter__(self):
        return self

    async def __aexit__(self, exc_type, exc_val, exc_tb):
        await self.kill()